lib/puppet/type/cisco_upgrade.rb

```diff
@@ -0,0 +1,116 @@
+#
+# Manages the version of Cisco Image running on a device.
+#
+# Copyright (c) 2017 Cisco and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+Puppet::Type.newtype(:cisco_upgrade) do
+  @doc = "Manages the version of Cisco Image running on a device.
+
+  ```
+  cisco_upgrade {\"<instance_name>\":
+    ..attributes..
+  }
+  ```
+
+  There can only be one instance of cisco_upgrade i.e. 'image'
+
+  Example:
+  ```
+    cisco_upgrade {'image' :
+      version           => '7.0(3)I5(1)',
+      source_uri        => 'bootflash:///nxos.7.0.3.I5.1.bin',
+      force_upgrade     => false,
+      delete_boot_image => false,
+    }
+  ```
+  "
+
+  # Parse out the title to fill in the attributes in these
+  # patterns. These attributes can be overwritten later.
+  def self.title_patterns
+    identity = ->(x) { x }
+    patterns = []
+
+    patterns << [
+      /^(\S+)$/,
+      [
+        [:name, identity]
+      ],
+    ]
+    patterns
+  end
+
+  newparam(:name, namevar: :true) do
+    # Parameter used only to satisfy namevar
+    desc 'Name of cisco_upgrade instance. Valid values are string'
+    validate do |name|
+      warning "only 'image' is accepted as a valid name" if name != 'image'
+    end
+  end
+
+  newparam(:source_uri) do
```

Comment:
Just realized that we need to make `:source_uri` mandatory.  If the manifest does not specify the parameter or if the `puppet resource cisco_upgrade image version=ver` command does not specify this parameter we get a very cryptic error.

Use a validate block at the end of the type to do a nil check and fail indicating the parameter is mandatory:

Example: https://github.com/cisco/cisco-network-puppet-module/blob/2f1d60e358b7009bfedcf6175357603c9e115831/lib/puppet/type/cisco_bgp_neighbor_af.rb#L490